### PR TITLE
Only output color esc. sequences if stdout is interactive

### LINKED ISSUE #7824: Color escape sequences emitted when stdout is not interactive term
We should check stdout for (terminal) interactivity when turning color on.

### PR DESCRIPTION
**Pull request by rwcarlsen**
Fixes #7824

### CODE REVIEW

**Comment by dschwen** [comment]
IDK, I pipe into `less -R` all the time. Why you want to take color from me?

**Comment by friedmud** [comment]
How does this work with the testing system?  I definitely want to see
coloring when tests fail or when using verbose mode.

(Honest question, it might still work...)

**Comment by dschwen** [comment]
I guess we'll see in a few minutes :-)

**Comment by rwcarlsen** [comment]
Would a ``--force-color`` flag be enough - or would there still be large angry hoards?

**Comment by friedmud** [comment]
Heh - I don't know about angry hordes... but as long as color comes through
the testing system I'll be good :-)

I agree with being conservative on coloring - always annoying when you get
a blast of escape characters and you didn't want them!

**Comment by dschwen** [comment]
In regular testing outputs you will not see a change. That color comes from the python test harness. Only if a test fails and output is printed (or if the tests are run with verbose output) you'd see a change. The `--force-color` parameter is fine with me. It should always be supplied by the test harness.

File: framework/include/base/MooseError.h [hunk 1]
```diff
@@ -42,10 +42,10 @@
   {                                                                                 \
     std::ostringstream _error_oss_;                                                 \
     _error_oss_ << "\n\n"                                                           \
-                << (Moose::_color_console ? XTERM_RED : "")                         \
+                << (Moose::colorConsole() ? XTERM_RED : "")                         \
```
Comment (dschwen) [inline framework/include/base/MooseError.h]: Why doesn't this just use the `COLOR_RED` macro defined above? (same below)

File: framework/src/outputs/DOFMapOutput.C [hunk 2]
```diff
@@ -53,14 +53,14 @@ DOFMapOutput::DOFMapOutput(const InputParameters & parameters) :
...
-  if (Moose::_color_console)
+  if (Moose::colorConsole())
   {
     char * term_env = getenv("TERM");
     if (term_env)
     {
       std::string term(term_env);
       if (term != "xterm-256color" && term != "xterm")
-        Moose::_color_console = false;
+        Moose::setColorConsole(false);
```
Comment (dschwen) [inline framework/src/outputs/DOFMapOutput.C]: Uhm, yeah... I wrote the DOFMap output object, but I have no idea why I thought it'd be a good idea to have an individual output object mess with the global output coloring flag. This was probably copy pasted from `Console`. This output object should just write to files anyways! Just remove this entire if block.

File: framework/include/base/Moose.h [hunk 3]
```diff
@@ -103,15 +98,24 @@ extern bool _throw_on_error;
...
+#define COLOR_BLUE    (Moose::colorConsole() ? XTERM_BLUE : "")
+#define COLOR_MAGENTA (Moose::colorConsole() ? XTERM_MAGENTA : "")
+#define COLOR_CYAN    (Moose::colorConsole() ? XTERM_CYAN : "")
+#define COLOR_WHITE   (Moose::colorConsole() ? XTERM_WHITE : "")
+#define COLOR_DEFAULT (Moose::colorConsole() ? XTERM_DEFAULT : "")
+
+
+
+ /// Returns whether Console coloring is turned on (default: true).
+ bool colorConsole();
```
Comment (rwcarlsen) [inline framework/include/base/Moose.h]: fixed

File: framework/src/outputs/DOFMapOutput.C (hunk 2 above)
Comment (rwcarlsen) [inline framework/src/outputs/DOFMapOutput.C]: removed.

File: framework/include/base/MooseError.h (hunk 1 above)
Comment (rwcarlsen) [inline framework/include/base/MooseError.h]: Updated entire file.

**Comment by rwcarlsen** [comment]
Switched to `--color=[auto,on,off]`.  Added deprecation warning to `--no-color` flag.

File: framework/src/base/MooseApp.C [hunk 4]
```diff
@@ -232,7 +233,8 @@ MooseApp::setupOptions()
     Moose::_deprecated_is_error = false;
 
   // Toggle the color console off
-  Moose::_color_console = !getParam<bool>("no_color");
+  Moose::setColorConsole(!getParam<bool>("no_color"));
+  Moose::setColorConsole(true, getParam<bool>("force_color"));
```
Comment (rwcarlsen) [inline framework/src/base/MooseApp.C]: deprec warn added

**Comment by rwcarlsen** [comment]
This should be ready for a final look/decision.

**Comment by rwcarlsen** [comment]
FWIW, run_tests and the test harness already strip out coloring (before this PR) of moose app output.

File: framework/src/outputs/DOFMapOutput.C [hunk 5]
```diff
@@ -53,15 +53,9 @@ DOFMapOutput::DOFMapOutput(const InputParameters & parameters) :
     _mesh(_problem_ptr->mesh())
 {
   // Set output coloring
-  if (Moose::_color_console)
+  if (Moose::colorConsole())
```
Comment (dschwen) [inline framework/src/outputs/DOFMapOutput.C]: this entire thing can go! the stuff you left does nothing anymore anyways

**Comment by friedmud** [comment]
> FWIW, run_tests and the test harness already strip out coloring (before this PR) of moose app output.

Not to my eyes:

color

File: framework/src/outputs/DOFMapOutput.C (hunk 5 above)
Comment (rwcarlsen) [inline framework/src/outputs/DOFMapOutput.C]: fixed.

**Comment by rwcarlsen** [comment]
@friedmud - weird - the coloring for errors is gone (before this PR) though.  I modded  RunApp.py to pass down ``--color on`` unless the pre-existing test harness --no-color flag is given.

**Comment by permcody** [comment]
The color problem isn't all that weird. I've given some thought to it and believe I know what the issue is. I've explained how this can happen here: https://github.com/idaholab/moose/issues/7735

File: python/TestHarness/testers/RunApp.py [hunk 6]
```diff
@@ -114,7 +114,9 @@ def getCommand(self, options):
       specs['cli_args'].append('Outputs/print_perf_log=true')
 
     if options.colored == False:
-      specs['cli_args'].append('--no-color')
+      specs['cli_args'].append('--color off')
+    else:
+      specs['cli_args'].append('--color on')
```
Comment (permcody) [inline python/TestHarness/testers/RunApp.py]: Do we need the condition in the else clause? Shouldn't the default be "auto" in the case that no flag is passed through the `TestHarness`?

**Review by permcody — COMMENTED** [review]
This looks all OK to me, it's a bit of a change for people who are used to that flag. Is there anyone opposed to this before it's merged?
@idaholab/moose-team

File: python/TestHarness/testers/RunApp.py (hunk 6 above)
Comment (rwcarlsen) [inline python/TestHarness/testers/RunApp.py]: @dschwen is correct - to preserve prev/normal test color behavior, we now need to force color on for tests.  Auto would be off.

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
Unfortunately this PR now causes all MPI runs to have no color 😦 .

**Comment by friedmud** [comment]
Yikes!

**Comment by friedmud** [comment]
That's a pretty big bug!

**Comment by permcody** [comment]
I've been down this path before... MPI destroys your ability to query anything from the environment. I had code to detect the terminal width to properly size the post processor table, only worked in serial (technically it only works if you run your executable without "mpiexec"). 

@rwcarlsen - I"m not sure there's a solution to this at all. I think we are screwed with trying to get any information out of the environment. I don't think I like the behavior of losing coloring in parallel.

**Pull request merged**